### PR TITLE
feat(FN-980): list previous reports backend

### PR DESCRIPTION
**Pull request by francescastocco**
## Introduction
We need to allow users to see a list of previously submitted reports.

## Resolution
- Endpoint in DTFS Central API to get DB collection `utilisation-reports`
- Endpoint in Portal API which calls above endpoint and maps data to format required in the frontend
- In Portal call above endpoint
- Started writing some tests but would be easier with the endpoint to create documents in this collection

## To Do
- Finish off tests
- Add some validation for bank ID

### CODE REVIEW

File: portal/server/api.js [hunk 1]
```diff
@@ -797,6 +797,24 @@ const getUkefDecisionReport = async (token, payload) => {
   }
 };
 
+const getPreviousUtilisationReportsByBank = async (token, bankId) => {
+  try {
+    const response = await axios({
+      method: 'get',
+      url: `${PORTAL_API_URL}/v1/previous-reports/${bankId}`,
```
Comment (ChristianMcc) [inline portal/server/api.js:804]: Same here for some extra validation on the bankID to check it's the right format

File: portal-api/src/v1/controllers/utilisation-report-service/previous-reports.controller.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  return groupedReports;
+};
+
+const getPreviousReportsByBankId = async (req, res) => {
+  try {
+    const { bankId } = req.params;
+
+    // Add validation for user bank ID here?
```
Comment (ChristianMcc) [inline portal-api/src/v1/controllers/utilisation-report-service/previous-reports.controller.js:66]: Can see you've added a comment here so ye just chuck it in a route validation and then in the api files

File: dtfs-central-api/src/v1/routes/portal-routes.js [hunk 3]
```diff
@@ -870,4 +872,6 @@ portalRouter.route('/cron-jobs').delete(cronJobsController.deleteAllEstoreLogs);
 portalRouter.route('/gef/mandatory-criteria/latest').get(mandatoryCriteria.getLatestGefMandatoryCriteria);
 portalRouter.route('/gef/mandatory-criteria/version/:version').get(mandatoryCriteria.getGefMandatoryCriteriaByVersion);
 
+portalRouter.route('/previous-reports/:bankId').get(previousReports.getUtilisationReports);
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/routes/portal-routes.js:875]: Can we add a validation on the bankId here too? it'll be the same as in portal API just checking the format of it

File: portal-api/src/v1/routes.js [hunk 4]
```diff
@@ -205,4 +207,6 @@ authRouter.get('/validate', (req, res) => {
 // bank-validator
 authRouter.get('/validate/bank', (req, res) => banks.validateBank(req, res));
 
+authRouter.route('/previous-reports/:bankId').get(validateUserHasAtLeastOneAllowedRole({ allowedRoles: [PAYMENT_OFFICER] }), getPreviousReportsByBankId);
```
Comment (ChristianMcc) [inline portal-api/src/v1/routes.js:210]: It's missing  from the rest of this file like in the other route but we should validate bankId here with the validator package just as in central API. I think TFM API has an example of how we can write some validations there

File: dtfs-central-api/src/v1/controllers/portal/utilisation-report-service/previous-reports.controller.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      $sort: {
+        year: 1,
+        month: 1,
+      },
+    },
+  ]).toArray();
+  if (filteredAndSortedUtilisationReports) {
+    return res.status(200).send(filteredAndSortedUtilisationReports);
+  }
+  return res.status(404).send({ status: 404, message: 'No utilisation reports found' });
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/controllers/portal/utilisation-report-service/previous-reports.controller.js:23]: I'm not sure 404 is right here, 404 to me is we've hit a route that doesn't exist.
If there are no reports isfilteredAndSortedUtilisationReports `[]` or `null` or `undefined`? If [] then that will be truthy in javascript and we'll return the 200 and empty array which is fine.
If it's not [] then we can assume an error happened (do we need a try/catch?) and we can send back 500 and Error finding reports?

File: dtfs-central-api/src/v1/routes/portal-routes.js [hunk 6]
```diff
@@ -870,4 +872,6 @@
...
```
[the remark this answers is not in the document]
Comment (ChristianMcc) [reply dtfs-central-api/src/v1/routes/portal-routes.js:875]: Is there global rate-limiting on this that means this is fine? the DTFS team implemented some rate limiting a few weeks ago so we should check with them if we can ignore this or if we need to do anything :)

**Review by ChristianMcc — CHANGES_REQUESTED** [review]
Some initial thoughts, still going through the grouping logic in portal-API but will come back to it later :)

File: dtfs-central-api/src/v1/controllers/portal/utilisation-report-service/previous-reports.controller.js (hunk 5 above)
Comment (francescastocco) [reply dtfs-central-api/src/v1/controllers/portal/utilisation-report-service/previous-reports.controller.js:23]: Yes yes, good point!

File: dtfs-central-api/src/v1/routes/portal-routes.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (francescastocco) [reply dtfs-central-api/src/v1/routes/portal-routes.js:875]: Yeh the DTFS team did implement this, I will test locally but probably a false positive

File: portal-api/src/v1/controllers/utilisation-report-service/previous-reports.controller.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      month: getMonthName(report.month),
+      path: report.path
+    }));
+  return {
+    year,
+    reports: reportsByYear,
+  };
+});
+
+const populateOmittedYears = (reportsGroupedByYear, years) => {
```
Comment (ChristianMcc) [inline portal-api/src/v1/controllers/utilisation-report-service/previous-reports.controller.js:29]: Minor nit-pick, can this return the array back so we reassign to a new variable in the function below please? little more obvious than a side effect of changing the array passed in etc.

**Review by ChristianMcc — COMMENTED** [review]
Looks good to me, I have one nit-pick that if you fix quickly then I'll approve :)
Agreed adding in missing years is a bit annoying but think your curent solution here is a good one!

File: portal-api/src/v1/controllers/utilisation-report-service/previous-reports.controller.js (hunk 7 above)
Comment (francescastocco) [reply portal-api/src/v1/controllers/utilisation-report-service/previous-reports.controller.js:29]: On it

**Pull request merged**